diff --git a/dwmblocks.c b/dwmblocks.c
index a2be9de..6034097 100644
--- a/dwmblocks.c
+++ b/dwmblocks.c
@@ -2,6 +2,7 @@
 #include<stdlib.h>
 #include<stdio.h>
 #include<string.h>
+#include<time.h>
 #include<unistd.h>
 #include<signal.h>
 #include<sys/wait.h>
@@ -57,6 +58,18 @@ static char statusstr[2][STATUSLENGTH];
 static int statusContinue = 1;
 static int returnStatus = 0;
 
+int gcd(int a, int b)
+{
+	int temp;
+	while (b > 0){
+		temp = a % b;
+
+		a = b;
+		b = temp;
+	}
+	return a;
+}
+
 //opens process *cmd and stores output in *output
 void getcmd(const Block *block, char *output)
 {
@@ -170,15 +183,25 @@ void pstdout()
 
 void statusloop()
 {
+	struct timespec tosleep;
 	setupsignals();
-	int i = 0;
+	unsigned int i = 0, interval;
+    for(unsigned int i = 0; i < LENGTH(blocks); i++){
+        if(blocks[i].interval){
+            interval = gcd(blocks[i].interval, interval);
+        }
+    }
 	getcmds(-1);
 	while (1) {
+		tosleep.tv_sec = interval;
+		tosleep.tv_nsec = 0;
+		if (nanosleep(&tosleep, &tosleep) == -1) {
+			continue;
+		}
 		getcmds(i++);
 		writestatus();
 		if (!statusContinue)
 			break;
-		sleep(1.0);
 	}
 }
 
