diff --git a/dwmblocks.c b/dwmblocks.c
index adb1e11..4e3aa90 100644
--- a/dwmblocks.c
+++ b/dwmblocks.c
@@ -70,6 +70,9 @@ void getcmd(const Block *block, char *output)
 		return;
 	int i = strlen(block->icon);
 	fgets(statusptr+i, CMDLENGTH-i-delimLen, cmdf);
+	if (fgetc(cmdf) != EOF) {
+		strncpy(statusptr + CMDLENGTH - delimLen - 4, "...", 4);
+	}
 	i = strlen(statusptr);
 	//if block and command output are both not empty
 	if (i != 0) {
